jsonnet/telemeter/server/kubernetes/kubernetes.libsonnet

```diff
@@ -0,0 +1,167 @@
+      local container = k.apps.v1beta2.statefulSet.mixin.spec.template.spec.containersType;
+      local volume = k.apps.v1beta2.statefulSet.mixin.spec.template.spec.volumesType;
+      local containerPort = container.portsType;
+      local containerVolumeMount = container.volumeMountsType;
+      local containerEnv = container.envType;
+
+      local podLabels = { 'k8s-app': 'telemeter-server' };
+      local localMount = containerVolumeMount.new(localVolumeName, localMountPath);
+      local localVolume = volume.fromEmptyDir(localVolumeName, {});
+      local tlsMount = containerVolumeMount.new(tlsVolumeName, tlsMountPath);
+      local tlsVolume = volume.fromSecret(tlsVolumeName, tlsSecret);
+      local name = containerEnv.fromFieldPath('NAME', 'metadata.name');
+      local namespace = containerEnv.fromFieldPath('NAMESPACE', 'metadata.namespace');
+      local rhdURL = containerEnv.fromSecretRef('RHD_URL', secretName, 'rhd.url');
+      local rhdUsername = containerEnv.fromSecretRef('RHD_USERNAME', secretName, 'rhd.username');
+      local rhdPassword = containerEnv.fromSecretRef('RHD_PASSWORD', secretName, 'rhd.password');
+      local rhdClientID = containerEnv.fromSecretRef('RHD_CLIENT_ID', secretName, 'rhd.client_id');
+
+      local telemeterServer =
+        container.new('telemeter-server', $._config.imageRepos.telemeterServer + ':' + $._config.versions.telemeterServer) +
+        container.withCommand([
+          '/usr/bin/telemeter-server',
+          '--join=telemeter-cluster',
+          '--name=$(NAME)',
+          '--listen=0.0.0.0:8443',
+          '--listen-internal=0.0.0.0:8081',
+          '--listen-cluster=0.0.0.0:8082',
+          '--storage-dir=' + localMountPath,
+          '--shared-key=%s/tls.key' % tlsMountPath,
+          '--tls-key=%s/tls.key' % tlsMountPath,
+          '--tls-crt=%s/tls.crt' % tlsMountPath,
+          '--authorize=' + $._config.telemeterServer.authorizeURL,
+          '--authorize-issuer-url=$(RHD_URL)',
+          '--authorize-client-id=$(RHD_CLIENT_ID)',
+          '--authorize-username=$(RHD_USERNAME)',
+          '--authorize-password=$(RHD_PASSWORD)',
+        ]) +
+        container.withPorts([
+          containerPort.newNamed('external', externalPort),
+          containerPort.newNamed('internal', internalPort),
+          containerPort.newNamed('cluster', clusterPort),
+        ]) +
+        container.withVolumeMounts([tlsMount, localMount]) +
+        container.withEnv([rhdURL, rhdUsername, rhdPassword, rhdClientID]) + {
+          livenessProbe: {
+            httpGet: {
+              path: '/healthz',
+              port: externalPort,
+              scheme: 'HTTPS',
+            },
+          },
+          readinessProbe: {
+            httpGet: {
+              path: '/healthz/ready',
+              port: externalPort,
+              scheme: 'HTTPS',
+            },
+          },
+        };
+
+      statefulSet.new('telemeter-server', 3, [telemeterServer], {}, podLabels) +
```

Comment:
just for readability, can we mixin those settings to make those named-parameter-like? for the reader it is very unclear what those parameters mean.

Reply:
Unfortunately, the ksonnet library does not specify defaults in the `new` constructor functions. That means two that eliminating the constructor leaves us with two bad options where we replace one single constructor line with 6+ additional mixins:
1. replace the constructor with another constructor with empty values and then manually override each field that would have been set. This is more confusing and very verbose: 
    ```
    statefulSet.new('', 0, [], {}, {}) +
    statefulSet.mixin.metadata.withName(name) + 
    statefulSet.mixin.spec.withReplicas(replicas) +
    statefulSet.mixin.spec.template.spec.withContainers(containers) +
    statefulSet.mixin.spec.withVolumeClaimTemplates(volumeClaims) +
    statefulSet.mixin.spec.template.metadata.withLabels(podLabels)
    ```

2. replace the constructor with all of the mixins plus manually specify the API version and kind, since these are not exposed in ksonnet:
    ```
    {apiVersion: "apps/v1beta2"} + 
    {kind: "StatefulSet"} + 
    statefulSet.mixin.metadata.withName(name) + 
    statefulSet.mixin.spec.withReplicas(replicas) +
    statefulSet.mixin.spec.template.spec.withContainers(containers) +
    statefulSet.mixin.spec.withVolumeClaimTemplates(volumeClaims) +
    statefulSet.mixin.spec.template.metadata.withLabels(podLabels)
    ````

Having to manually specify the API version + kind is error prone and seems very one-off; and specifying the empty constructor just to get the correct basic mixin is confusing and harder to understand. For these reasons, I don't think it is a good idea to eschew the constructor. What do you think?

Reply:
I think it's enough to have something like `$._config.telemeterServer.replicas` and pass that one into the `.new()` function.

Reply:
@squat thanks a lot for the analysis! I was hoping the statefulset would have a non-param constructor like ingress does:

https://github.com/s-urbaniak/k8s-hetzner/blob/f849a24c32b8854bf3cfeaadc971025e6cd9b4a4/jsonnet/telemeter.jsonnet#L8-L14

This is no way is a stopper, it would be just nice to hint somehow the semantic meaning of those parameters.

Reply:
@s-urbaniak you got it. And to clarify: yes ingress and many of the more basic "building" blocks of K8s (pods, bindings, endpoints, etc) have ksonnet constructors that take 0 parameters, but the higher-level objects like Deployments and StatefulSets are parameterized. jsonnet allows for functions to have default values but this is not leveraged currently in constructors in ksonnet.